src/rsa.rs

```diff
@@ -163,8 +165,107 @@ fn parse_public_key<'a>(input: untrusted::Input<'a>) ->
     })
 }
 
+#[allow(non_camel_case_types)]
+enum BIGNUM {}
+#[allow(non_camel_case_types)]
+enum BN_MONT_CTX {}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct RSA {
+    n: *mut BIGNUM,
+    e: *mut BIGNUM,
+    d: *mut BIGNUM,
+    p: *mut BIGNUM,
+    q: *mut BIGNUM,
+    dmp1: *mut BIGNUM,
+    dmq1: *mut BIGNUM,
+    iqmp: *mut BIGNUM,
+    mont_n: *mut BN_MONT_CTX,
+    mont_p: *mut BN_MONT_CTX,
+    mont_q: *mut BN_MONT_CTX,
+    mont_qq: *mut BN_MONT_CTX,
+    qmn_mont: *mut BIGNUM,
+    iqmp_mont: *mut BIGNUM,
+}
+
+/// An RSA key pair.
+pub struct RSAKeyPair {
+    rsa: Box<RSA>,
+}
+
+const BN_FLG_CONSTTIME: c::int = 4;
+
+fn build_bn(input: untrusted::Input) -> *mut BIGNUM {
+    unsafe {
+        let bytes = input.as_slice_less_safe();
+        let res = BN_bin2bn(bytes.as_ptr(), bytes.len(), ptr::null_mut());
+        BN_set_flags(res, BN_FLG_CONSTTIME);
+        res
+    }
```

Comment:
See https://github.com/briansmith/ring/blob/master/STYLE.md#unsafe, in particular the part about minimizing the scope of the `unsafe`. In particular, `input.as_slice_less.safe()` shouldn't be inside the `unsafe`.

Do you need to use `BN_set_flags` here? Won't `rsa_new_end` set it on everything as needed?

Let's consider the error handling strategy. BN_bin2bn might return a null pointer when it fails. It seems you want to return the NULL pointer. Then, if any of the `BN_bin2bn` calls fail, we'll construct an `RSA` that has at least one NULL pointer member. Then, we are expecting `rsa_new_end` to check for all these NULL pointers. But, that seems error prone, and in fact `rsa_new_end` doesn't check for the NULL pointers (it only `asserts` they aren't NULL). Further, this seems prone to leak the `BIGNUM`, if we were to return early before constructing the `RSA`.

Maybe it would make more sense for this function to return `Result<PositiveInteger, ()>`, where `struct PositiveInteger { value: *mut BIGNUM }`, where PositiveInteger works exactly like `Box` except that it uses `BN_free()` instead of `free()`? Then, when we construct the `RSA`, we'd do it like `Box::new(RSA { n: n.into_raw(), e: e.into_raw(), ... })`.

Note that `build_bn` (perhaps renamed `PositiveInteger::new`) could also be the thing that calls `der::positive_integer`.